### PR TITLE
Make DisplayableException more serialization-friendly

### PR DESCRIPTION
**Pull request by minestarks**
Fixes https://dev.azure.com/ms-quantum/Quantum%20Program/_workitems/edit/44472

Serializing `DisplayableException` to JSON causes an `OutOfMemoryException` to be thrown. 

JSON tracing shows hundreds of objects referenced through `StackFrame` attempted to be serialized:

```
Newtonsoft.Json Information: 0 : Finished serializing System.RuntimeType. Path 'StackTrace[0].Callable.__Body__.Method.Module.Assembly.DefinedTypes[263]'.
Newtonsoft.Json Information: 0 : Started serializing System.RuntimeType. Path 'StackTrace[0].Callable.__Body__.Method.Module.Assembly.DefinedTypes[263]'.
Newtonsoft.Json Information: 0 : Finished serializing System.RuntimeType. Path 'StackTrace[0].Callable.__Body__.Method.Module.Assembly.DefinedTypes[264]'.
Newtonsoft.Json Information: 0 : Started serializing System.RuntimeType. Path 'StackTrace[0].Callable.__Body__.Method.Module.Assembly.DefinedTypes[264]'.
```

On principle I don't think an object designed to be serialized/displayed should have references to complicated types such as `StackFrame`. So I'm introducing a `DisplayableStackFrame`, a simple struct.

Additionally I'm cleaning up a few areas that I thought looked odd - where we were serializing the same string slightly differently, unnecessary null handling, etc.

Considered, and rejected:
 - Adding a custom JSON converter for `DisplayableException`: designing the object to be simple in the first place seems like a more sensible approach
 - Avoid serializing to JSON in the first place: This turned out to be way trickier than it sounds. Serializing results to JSON seems to be pretty baked into the IQ# engine, and the JSON results are used by the Python wrapper (namely in [`capture_diagnostics`](https://github.com/microsoft/iqsharp/blob/d90578bb70907aa9bdcb4e1234c13f19ea0e27dc/src/Python/qsharp-core/qsharp/clients/iqsharp.py#L225))

### CODE REVIEW

File: src/Jupyter/Extensions.cs [hunk 1]
```diff
@@ -103,11 +103,7 @@ public static T WithStackTraceDisplay<T>(this T simulator, IChannel channel)
         simulator.DisableExceptionPrinting();
         simulator.OnException += (exception, stackTrace) =>
         {
-            channel.Display(new DisplayableException
-            {
-                Exception = exception,
-                StackTrace = stackTrace
-            });
+            channel.Display(DisplayableException.Create(exception, stackTrace));
```
Comment (minestarks) [inline src/Jupyter/Extensions.cs:106]: The `DisplayableException` here will end up [being serialized](https://github.com/microsoft/jupyter-core/blob/4d9233bdede5c0b7135d4e0fc8e4ecb24a54ab4d/src/Engines/BaseEngine.cs#L357) into various formats (text, html and JSON) to be included in a message to sent back to the Jupyter frontend.

File: src/Jupyter/Visualization/DisplayableExceptionEncoders.cs [hunk 2]
```diff
@@ -2,27 +2,38 @@
...
 namespace Microsoft.Quantum.IQSharp
 {
-    internal record struct DisplayableException(Exception Exception, IEnumerable<StackFrame>? StackTrace)
+    internal record struct DisplayableStackFrame(string Callable, string SourceFile, string BestSourceLocation, int LineNumber)
     {
+    }
+
+    internal record struct DisplayableException(string? ExceptionType, string ExceptionMessage, IEnumerable<DisplayableStackFrame> StackTrace)
+    {
+        public static DisplayableException Create(Exception exception, IEnumerable<StackFrame> stackTrace)
```
Comment (minestarks) [inline src/Jupyter/Visualization/DisplayableExceptionEncoders.cs:19]: The problematic type is [`StackFrame`](https://github.com/microsoft/qsharp-runtime/blob/09403e3d306567f2cff237881dbb064829b496a5/src/Simulation/Simulators/CommonNativeSimulator/StackTrace.cs#L17), which has a reference to `Callable`, which can have references to `Assembly` objects, etc., etc., which ultimately fan out into hundreds of types to be serialized. We shouldn't have this type in a class that's intended to be serialized.

File: src/Jupyter/Visualization/DisplayableExceptionEncoders.cs [hunk 3]
```diff
@@ -74,10 +90,12 @@ public class DisplayableExceptionToHtmlEncoder : IResultEncoder
                         {String.Join("\n", rows)}
                     </tbody>
                 ";
+
+                // { display: list-item } style is needed for a <summary> tag so that the arrow shows up
```
Comment (minestarks) [inline src/Jupyter/Visualization/DisplayableExceptionEncoders.cs:94]: Jupyter's stylesheets (I think) were causing the disclosure triangle to be hidden, so it wasn't clear you could click on the element to see the callstack.

With this change, it should look like:

image

See: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/summary#default_style

File: src/Jupyter/Extensions.cs (hunk 1 above)
Comment (minestarks) [reply src/Jupyter/Extensions.cs:106]: Jupyter only displays the HTML output, but the Python wrapper uses [JSON results](https://github.com/microsoft/iqsharp/blob/d90578bb70907aa9bdcb4e1234c13f19ea0e27dc/src/Python/qsharp-core/qsharp/clients/iqsharp.py#L388), which is why I can't just simply disable JSON serialization for this type.

File: src/Jupyter/Visualization/DisplayableExceptionEncoders.cs [hunk 4]
```diff
@@ -35,10 +46,10 @@ internal static string ToFriendlyName(this ICallable callable)
             : fullName;
         }
 
-        internal static string ToSourceLink(this StackFrame frame) =>
```
Comment (DmitryVasilevsky) [inline src/Jupyter/Visualization/DisplayableExceptionEncoders.cs:38]: Should this be an extension method or could it be moved to the class?

File: src/Jupyter/Visualization/DisplayableExceptionEncoders.cs [hunk 5]
```diff
@@ -54,15 +65,20 @@ public class DisplayableExceptionToHtmlEncoder : IResultEncoder
...
+                // upon the failure event raised by the simulator
+                System.Diagnostics.Debug.Assert(ex.StackTrace != null);
+
                 var rows = ex
                     .StackTrace
-                    ?.Select(frame => $@"
+                    .Select(frame => $@"
                         <tr>
                             <td>{frame.ToSourceLink()}</td>
-                            <td>{frame.Callable.ToFriendlyName()}</td>
```
Comment (DmitryVasilevsky) [inline src/Jupyter/Visualization/DisplayableExceptionEncoders.cs:62]: Do we still need .ToFriendlyName() extension method on ICallable?

File: src/Jupyter/Visualization/DisplayableExceptionEncoders.cs [hunk 6]
```diff
@@ -74,10 +90,12 @@ public class DisplayableExceptionToHtmlEncoder : IResultEncoder
...
                     </tbody>
                 ";
+
+                // { display: list-item } style is needed for a <summary> tag so that the arrow shows up 
                 var table = $@"
                     <details>
-                        <summary>
-                            Unhandled exception of type {ex.Exception.GetType().FullName}: {ex.Exception.Message}
+                        <summary style=""display:list-item"">
+                            {ex.Header}
```
Comment (DmitryVasilevsky) [inline src/Jupyter/Visualization/DisplayableExceptionEncoders.cs:98]: Should it be HTML Encoded?

File: src/Jupyter/Visualization/DisplayableExceptionEncoders.cs [hunk 7]
```diff
@@ -107,11 +125,11 @@ public class DisplayableExceptionToTextEncoder : IResultEncoder
                 var builder = new StringBuilder();
                 builder.AppendLine(ex.Header);
                 var first = true;
-                foreach (var frame in ex.StackTrace ?? Enumerable.Empty<StackFrame>())
+                foreach (var frame in ex.StackTrace ?? Enumerable.Empty<DisplayableStackFrame>())
```
Comment (DmitryVasilevsky) [inline src/Jupyter/Visualization/DisplayableExceptionEncoders.cs:128]: In the code above ex.StackTrace is accessed without the null-coalescing operator (??). Do we need it here?

File: src/Jupyter/Visualization/DisplayableExceptionEncoders.cs (hunk 4 above)
Comment (minestarks) [reply src/Jupyter/Visualization/DisplayableExceptionEncoders.cs:38]: Good point, no reason for the extension class anymore

File: src/Jupyter/Visualization/DisplayableExceptionEncoders.cs (hunk 7 above)
Comment (minestarks) [reply src/Jupyter/Visualization/DisplayableExceptionEncoders.cs:128]: No, good point, the assumptions should be the same.

**Review by cesarzc — APPROVED** [review]

**Pull request merged**